[CodeGen][arm64e] Add methods and data members to Address, which are needed to authenticate signed pointers

ahatanak opened the pull request:
To authenticate pointers, CodeGen needs access to the key and discriminators that were used to sign the pointer. That information is sometimes known from the context, but not always, which is why `Address` needs to hold that information.

This patch adds methods and data members to `Address`, which will be needed in subsequent patches to authenticate signed pointers, and uses the newly added methods throughout CodeGen. Although this patch isn't strictly NFC as it causes CodeGen to use different code paths in some cases (e.g., `mergeAddressesInConditionalExpr`), it doesn't cause any changes in functionality as it doesn't add any information needed for authentication.

In addition to the changes mentioned above, this patch introduces class `RawAddress`, which contains a pointer that we know is unsigned, and adds several new functions for creating `Address` and `LValue` objects.

ahatanak committed:
[clang][CodeGen] Introduce class RawAddress and modify Address in
preparation for adding information to it that is needed for pointer
authentication

fhahn commented on clang/lib/CodeGen/Address.h:
```diff
@@ -15,35 +15,42 @@
 
-/// An aligned address.
-class Address {
+class RawAddress {
```
Would be good to have a brief comment here documenting `RawAddress`

fhahn reviewed (commented):
It might be helpful for reviewers to add a brief  description to the PR explaining the motivation and implementation

efriedma-quic commented on clang/lib/CodeGen/Address.h:
```diff
@@ -80,6 +94,139 @@ class Address {
+  /// this is currently redundant with the pointer's type, but for signed
+  /// pointers it is useful if the pointer has been offsetted or cast from the
+  /// original type. In the long run, when LLVM adopts opaque pointer types,
+  /// this should become the notional element type of the address.
```
This comment probably needs to be updated?

efriedma-quic commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -232,19 +232,19 @@ static Value *MakeBinaryAtomicValue(
+  Address Addr = CGF.EmitPointerWithAlignment(E->getArg(0));
 
-  LValue LV = CGF.MakeNaturalAlignAddrLValue(Address, E->getType());
+  LValue LV = CGF.MakeAddrLValue(Addr, E->getType());
```
Can we land this separately with appropriate tests?  It looks like it changes the behavior (the computed alignment of a pointer can be less than natural alignment, I think).

ahatanak replied:
My thinking was that the call to `EmitPointerWithAlignment`, which sets `Addr`, would compute the correct alignment.

Do you have an example that shows the alignment is lower after this change?

efriedma-quic replied:
EmitPointerWithAlignment tries to compute the alignment based on the underlying lvalue.  This can be higher or lower than the natural alignment of the type.  Say you have something like `vec f() { struct S { char c[16]; } x; return __temporal_load((vec*)x.c); }`.  It looks through the cast, sees the field is unaligned, and therefore concludes the pointer is unaligned.  This is arguably an improvement, but it's a significant change to the generated code.

rjmccall replied:
Yeah, I agree that (1) this is an improvement and (2) we should probably isolate all of these alignment changes as separate patches.

ahatanak replied:
The patch that fixes the alignment: https://github.com/llvm/llvm-project/pull/75675

efriedma-quic commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -395,27 +395,27 @@ namespace {
-  // Deactivate the cleanup block.
-  DeactivateCleanupBlock(cleanup,
-                         cast<llvm::Instruction>(typedAddr.getPointer()));
+    pushFullExprCleanup<FreeException>(EHCleanup, addr.getRawPointer(*this));
```
The new indentation here doesn't look right?

asl commented on clang/lib/CodeGen/CGBuilder.h:
```diff
@@ -232,110 +279,133 @@ class CGBuilderTy : public CGBuilderBaseTy {
-    return CreateMemCpy(Dest.getPointer(), Dest.getAlignment().getAsAlign(),
-                        Src.getPointer(), Src.getAlignment().getAsAlign(), Size,
-                        IsVolatile);
+    return CreateMemCpy(getRawPointerFromAddress(Dest),
```
Here we are having a subtle, but big problem. With pauth,`getRawPointerFromAddress` has a side effect. However, the argument evaluation order is unspecified. As a result, the side effects could appear in arbitrary order.

See https://github.com/access-softek/llvm-project/issues/51 as an outcome of this, when pauth codegen is different on Mac and Linux.

rjmccall replied:
I agree this needs to be fixed.

asl reviewed (changes requested):
See the comment about function argument evaluation order that causes instability of the codegen with pauth.

ahatanak committed:
Address review comments

- Fix indentation.
- Update comment for Address:ElementType.
- Fix a bug where calls to getRawPointer in call argument expressions
  were causing the output to be dependent on the compiler.

ahatanak committed:
Fix indentation

ahatanak commented:
I think I've addressed all the feedback I got. Are there any other comments?

ahatanak commented:
ping

asl commented:
I am ok with the changes. Though with pauth we'd probably need to rename `getRawPointerFromAddress` to indicate clearer that it has side effects (like `extractRawPointerFromAddress` or something like this)

asl commented:
We discussed the naming thing (`extractRawPointerFromAddress` vs `getRawPointerFromAddress`) at pauth call and decided that it would be better to do this renaming now. @ahmedbougacha will coordinate that renaming.

ahatanak commented:
I'm not sure `extractRawPointerFromAddress` conveys the fact that the function might do code-gen instead of just returning some pointer. I wonder if there's a better name.

`computeRawPointerFromAddress`
`genRawPointerFromAddress`
`generateRawPointerFromAddress`
`codeGenRawPointerFromAddress`

Thoughts?

asl commented:
> I'm not sure `extractRawPointerFromAddress` conveys the fact that the function might do code-gen instead of just returning some pointer. I wonder if there's a better name.
> 
> `computeRawPointerFromAddress` `genRawPointerFromAddress` `generateRawPointerFromAddress` `codeGenRawPointerFromAddress`
> 
> Thoughts?

I do not have particular preference. But probably like `compute` / `gen` slightly more.

ahatanak commented:
Maybe `emitRawPointerFromAddress` is better. I see a lot of functions starting with `emit` in CodeGen.

asl commented:
> Maybe `emitRawPointerFromAddress` is better. I see a lot of functions starting with `emit` in CodeGen.

Works for me!

ahatanak committed:
Rename getRawPointer/getRawPointerFromAddress to emitRawPointer/emitRawPointerFromAddress

ahatanak commented:
Any other comments? Do the changes look good?
